_posts/2024-01-18-fido-passkey.md

```diff
@@ -0,0 +1,1194 @@
+
+
+We need to save the following data for the passkey.
+
+The key id is a unique identifier assigned to a registered key and is used to identify that specific key.\
+The public key is the user's public key generated during the registration process with an authenticator.\
+The signature count refers to a counter indicating the number of signatures with a specific FIDO authenticator. With this value, the server can verify if the counter has been correctly updated compared to the previous value. If the value is not correctly updated, it may indicate the possibility of a replay or an attack.\
+Transport refers to the manner in which communication occurs between the FIDO client (for example, an authenticator like a USB security key) and the FIDO server during the authentication process. The transport mechanism determines how data is exchanged between the authenticator and the server.\
+The type refers to the type of authenticator used to perform FIDO-based authentication.
+
+There are various transport mechanisms defined in FIDO, including:
+
+* USB: The authenticator is physically connected to the device through a USB port.
+* NFC (Near Field Communication): The authenticator and the device communicate wirelessly via NFC technology, which is convenient for mobile devices.
+* Bluetooth: Communication between the authenticator and the device occurs wirelessly via Bluetooth.
+* Internal: The authenticator is embedded within the device itself, such as a built-in fingerprint reader in a laptop.
+* Platform: Communication takes place through the platform itself, for example, between a mobile app on a smartphone and the authenticator.
+
+Create a SQL file `V1__init.sql`.
+```sql
+CREATE TABLE user
+(
+    id                      CHAR(36)                NOT NULL,
+    username                CHARACTER VARYING(255)  NOT NULL,
+    user_handle             VARBINARY(255)          NOT NULL,
+    assertion               MEDIUMTEXT,
+    public_key_json         MEDIUMTEXT,
+    registration_complete   BIT                     NOT NULL,
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE passkey
+(
+    id              CHAR(36)                NOT NULL,
+    key_id          VARBINARY(255)          NOT NULL,
+    public_key      VARBINARY(255)          NOT NULL,
+    signature_count INTEGER(20)             NOT NULL,
+    transport       CHARACTER VARYING(255)  NOT NULL,
+    type            CHARACTER VARYING(255)  NOT NULL,
+    user_handle     VARBINARY(255)          NOT NULL,
+    PRIMARY KEY (id)
+);
+```
+
+And configure the h2 database in the `application.properties` file.
+```properties
+spring.datasource.url=jdbc:h2:mem:db;DB_CLOSE_DELAY=-1;MODE=MySQL;NON_KEYWORDS=USER
+spring.datasource.driverClassName=org.h2.Driver
+```
+
+### Create utils
+Because Yubico's dependency works with their own object ByteArray instead of byte[], we create a small class to easily convert between them.
```

Comment:
```suggestion
### Create utils

Because Yubico's dependency works with their own object ByteArray instead of byte[], we create a small class to easily convert between them.
```